examples/ArduinoEthernetShieldSleep/ArduinoEthernetShieldSleep.ino

```diff
@@ -0,0 +1,203 @@
+ * 
+ *  
+ *  by Jose Manuel Perez
+ *  https://github.com/jmpmscorp *  
+ */
+
+#include <Ethernet.h>
+#include <utility/w5100.h>
+#include <MQTT.h>
+
+#include <Wire.h>
+#include "RTClib.h"   // From https://github.com/adafruit/RTClib
+
+#include <avr/sleep.h>
+#include <avr/wdt.h>
+
+byte mac[] = {0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0xED};
+byte ip[] = {192, 168, 1, 66};  // <- change to match your network
+
+const char * mqttBrokerAddress = "192.168.1.65";
+const char * mqttClientId = "arduino-ethernet";
+const char * mqttUsername = "";
+const char * mqttPassword = "";
+
+EthernetClient net;
+MQTTClient client;
+RTC_DS1307 rtc;
+
+volatile bool ethShieldInterruptFlag = false;
+volatile bool wdtInterruptFlag = false;
+bool didWakeup = true;
+
+unsigned long mqttLastActivityMillis = 0;
+
+void connect() {
+  Serial.print("connecting...");
+  while (!client.connect(mqttClientId, mqttUsername, mqttPassword)) {
+    Serial.print(".");
+    delay(1000);
+  }
+
+  Serial.println("\nconnected!");
+
+  client.subscribe("test/receive");
+}
+
+void messageReceived(String &topic, String &payload) {
+  Serial.println("incoming: " + topic + " - " + payload);
+}
+
+void setup() {
+  wdt_disable();
+  Serial.begin(19200);
+  Ethernet.begin(mac, ip);
+  while(!rtc.begin()) {
+    Serial.println(F("Couldn't find RTC"));
+    delay(2500);
+  }
+
+  rtc.adjust(DateTime(F(__DATE__), F(__TIME__)));
+
+  Serial.println(F("Start!"));
+  
+  // Note: Local domain names (e.g. "Computer.local" on OSX) are not supported by Arduino.
+  // You need to set the IP address directly.
+  client.begin(mqttBrokerAddress, net);
+  client.setClockSource(customMillis);
+  client.setOptions(30, true, 1500);
+  client.onMessage(messageReceived);
+  
+  // enable interrupts for Socket 0
+  W5100.writeIMR(0x01);
+  attachInterrupt(digitalPinToInterrupt(2), [](){ ethShieldInterruptFlag = true; }, FALLING );
+  
+  connect();
+  enableWdt(WDTO_8S);
+}
+
+void loop() {
+  if(wdtInterruptFlag) {
+    wdt_reset();
+    // Clear WDT Int Flag
+    WDTCSR |= _BV(WDIE);
+    wdtInterruptFlag = false;
+  }
+  
+  client.loop();
+
+  if (!client.connected()) {
+    connect();
+  }
+  else {
+    periodicallySent();
+  }
+
+  if(ethShieldInterruptFlag) {
+    ethShieldInterruptFlag = false;
+    //Serial.println("Ethernet Shield Interrupt");
+    uint8_t irState = W5100.readSnIR(0);
+    //Serial.println(irState, HEX);
+    W5100.writeSnIR(0, irState);
+    mqttLastActivityMillis = customMillis();
+  }
+
+  if(customMillis() - mqttLastActivityMillis > 2500) {
+    Serial.println(F("To Sleep!"));
+    Serial.flush();
+    enterSleep();
+  }
+}
+
+void periodicallySent() {
+  static unsigned long lastSentMillis = 0;
+
+  if(customMillis() - lastSentMillis > 60000) {
+    if(client.publish("test/period", "Periodically packet")) {
+      lastSentMillis = customMillis();
+    }    
+  }
+}
+
+uint32_t customMillis() {
+  static uint32_t offset = 0;
+
+  if(didWakeup) {
+    offset = rtcMillis() - millis();
```

Comment:
If you set `didWakeup = false` after setting the offset, `customMillis` will call `rtcMillis` only once after wakeup, right?

Reply:
Oh yes, of course! I forgot it. Feel free to add it.